allennlp/models/semantic_parsing/wikitables_variable_free/wikitables_variable_free_semantic_parser.py

```diff
@@ -0,0 +1,702 @@
+
+        if num_linking_features > 0:
+            self._linking_params = torch.nn.Linear(num_linking_features, 1)
+        else:
+            self._linking_params = None
+
+        if self._use_neighbor_similarity_for_linking:
+            self._question_entity_params = torch.nn.Linear(1, 1)
+            self._question_neighbor_params = torch.nn.Linear(1, 1)
+        else:
+            self._question_entity_params = None
+            self._question_neighbor_params = None
+
+    def _get_initial_rnn_and_grammar_state(self,
+                                           question: Dict[str, torch.LongTensor],
+                                           table: Dict[str, torch.LongTensor],
+                                           world: List[WikiTablesVariableFreeWorld],
+                                           actions: List[List[ProductionRuleArray]],
+                                           outputs: Dict[str, Any]) -> Tuple[List[RnnStatelet],
+                                                                             List[GrammarStatelet]]:
+        """
+        Encodes the question and table, computes a linking between the two, and constructs an
+        initial RnnStatelet and GrammarStatelet for each batch instance to pass to the
+        decoder.
+
+        We take ``outputs`` as a parameter here and `modify` it, adding things that we want to
+        visualize in a demo.
+        """
+        table_text = table['text']
+        # (batch_size, question_length, embedding_dim)
+        embedded_question = self._question_embedder(question)
+        question_mask = util.get_text_field_mask(question).float()
+        # (batch_size, num_entities, num_entity_tokens, embedding_dim)
+        embedded_table = self._question_embedder(table_text, num_wrapping_dims=1)
+        table_mask = util.get_text_field_mask(table_text, num_wrapping_dims=1).float()
+
+        batch_size, num_entities, num_entity_tokens, _ = embedded_table.size()
+        num_question_tokens = embedded_question.size(1)
+
+        # (batch_size, num_entities, embedding_dim)
+        encoded_table = self._entity_encoder(embedded_table, table_mask)
+
+        # entity_types: tensor with shape (batch_size, num_entities), where each entry is the
+        # entity's type id.
+        # entity_type_dict: Dict[int, int], mapping flattened_entity_index -> type_index
+        # These encode the same information, but for efficiency reasons later it's nice
+        # to have one version as a tensor and one that's accessible on the cpu.
+        entity_types, entity_type_dict = self._get_type_vector(world, num_entities, encoded_table)
+
+        entity_type_embeddings = self._entity_type_encoder_embedding(entity_types)
+
+        # (batch_size, num_entities, num_neighbors) or None
+        neighbor_indices = self._get_neighbor_indices(world, num_entities, encoded_table)
+
+        if neighbor_indices is not None:
```

Comment:
This part is new. It's possible that none of the instances in a given batch have entities with neighbors, because of the way the knowledge graph is defined.